TripalCultivate/TripalCultivate-Phenotypes
# Update feedback engine template

**Issue #114 **

## Motivation
This PR will enable the validation result window template to render failed items structured as a Drupal render array. 



## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an automated test to ensure it doesn't return.*

1. Modified the template to render failed item details render array on the fly.
2. Implemented a script that will inspect the DOM to apply specific class names to lengthy details. 
3. Applied visual cue element that partially reveals bottom item to indicate more items for scroll.
4. Implemented a kernel automated test.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

ClassName->testResultWindowDisplay:
- Test various details render array types: item, markup, list, list+table, raw row, image, link, concurrent items, and malformed render array.
- Test that validation item has the correct status-based class name.
- Test that validation item title matches the rendered markup output.
- Test the details rendered array rendered output can be found in the overall validation window markup output.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Checkout the branch g4.114-updateValidationFeedback
2. Setup Tripal Cultivate Phenotypes so that Traits Importer Page is accessible.
3. By pass the validation result window requiring a data file by adding the code snippet below in line number 352 of TripalCultivatePhenotypesTraitsImporter.php

Refer to this issue for details render array structures:
https://github.com/TripalCultivate/TripalCultivate-Phenotypes/issues/114#issuecomment-2486787283

```
// Full validation result summary.
if (isset($storage[$this->validation_result])) {
....  
}
// Line 352 here
 
// Begin line to add
$validation_result = [
  [
     'title' => 'Validation Title',
     'status' => 'pass',
     'details' => []
  ],

  // Another validation result item here...
];

$form['validation_result'] = [
  '#type' => 'inline_template',
  '#theme' => 'result_window',
  '#data' => [
    'validation_result' => $validation_result
  ],
  '#weight' => -100
];

```

4. Save, clear cache and reload page.

**NOTES:** 
- do not commit changes.
- test partial reveal and scrollbars with a detailed and lengthy failed item.
- test raw row by applying #attributes and set class to tcp-raw-row.
